internal/controller/postgrescluster/postgres.go

```diff
@@ -615,3 +602,102 @@ func (r *Reconciler) reconcilePostgresWALVolume(
 
 	return pvc, err
 }
+
+// reconcileDatabaseInitSQL runs custom SQL files in the database. When
+// DatabaseInitSQL is defined, the function will find the primary pod and run
+// SQL from the defined ConfigMap
+func (r *Reconciler) reconcileDatabaseInitSQL(ctx context.Context,
+	cluster *v1beta1.PostgresCluster, instances *observedInstances) error {
+	log := logging.FromContext(ctx)
+
+	// Spec is not defined, unset status and return
+	if cluster.Spec.DatabaseInitSQL == nil {
+		// If database init sql is not requested, we will always expect the
+		// status to be nil
+		cluster.Status.DatabaseInitSQL = nil
+		return nil
+	}
+
+	// Spec is defined but status is already set, return
+	if cluster.Status.DatabaseInitSQL != nil {
+		return nil
+	}
+
+	// Based on the previous checks, the user wants to run sql in the database.
+	// Check the provided ConfigMap name and key to ensure the a string
+	// exists in the ConfigMap data
+	var (
+		err  error
+		data string
+	)
+
+	getDataFromConfigMap := func() (error, string) {
+		cm := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      cluster.Spec.DatabaseInitSQL.Name,
+				Namespace: cluster.Namespace,
+			},
+		}
+		err := r.Client.Get(ctx, client.ObjectKeyFromObject(cm), cm)
+		if err != nil {
+			return err, ""
+		}
+
+		key := cluster.Spec.DatabaseInitSQL.Key
+		if _, ok := cm.Data[key]; !ok {
+			err := errors.Errorf("ConfigMap did not contain expected key: %s", key)
+			return err, ""
+		}
+
+		return nil, cm.Data[key]
+	}
+
+	if err, data = getDataFromConfigMap(); err != nil {
+		log.Error(err, "Could not get data from ConfigMap",
+			"ConfigMap", cluster.Spec.DatabaseInitSQL.Name,
+			"Key", cluster.Spec.DatabaseInitSQL.Key)
+		return err
```

Comment:
🌱 We should log an error or return it, not both. In the future, we want errors that are returned to become warning events on the cluster.

I picked up that mantra from thoughts like these:
- https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully
- https://dave.cheney.net/2015/11/05/lets-talk-about-logging